crates/webidl/src/lib.rs

```diff
@@ -96,60 +94,156 @@ trait WebidlParse<Ctx> {
-            webidl::ast::Definition::Interface(ref interface) => {
-                interface.webidl_parse(program, ())
+impl<'a, 'b> WebidlParse<&'a FirstPass<'b>> for webidl::ast::Definition {
+    fn webidl_parse(
+        &self,
+        program: &mut backend::ast::Program,
+        first_pass: &'a FirstPass<'b>,
+    ) -> Result<()> {
+        match self {
+            webidl::ast::Definition::Enum(enumeration) => enumeration.webidl_parse(program, ())?,
+            webidl::ast::Definition::Includes(includes) => {
+                includes.webidl_parse(program, first_pass)?
             }
-            webidl::ast::Definition::Typedef(ref typedef) => typedef.webidl_parse(program, ()),
-            webidl::ast::Definition::Enum(ref enumeration) => enumeration.webidl_parse(program, ()),
+            webidl::ast::Definition::Interface(interface) => {
+                interface.webidl_parse(program, first_pass)?
+            }
+            webidl::ast::Definition::Typedef(typedef) => typedef.webidl_parse(program, first_pass)?,
             // TODO
             webidl::ast::Definition::Callback(..)
             | webidl::ast::Definition::Dictionary(..)
             | webidl::ast::Definition::Implements(..)
-            | webidl::ast::Definition::Includes(..)
-            | webidl::ast::Definition::Mixin(..)
             | webidl::ast::Definition::Namespace(..) => {
-                warn!("Unsupported WebIDL definition: {:?}", self);
-                Ok(())
+                warn!("Unsupported WebIDL definition: {:?}", self)
+            }
+            webidl::ast::Definition::Mixin(_) => {
+                // handled in the first pass
             }
         }
+        Ok(())
     }
 }
 
-impl WebidlParse<()> for webidl::ast::Interface {
-    fn webidl_parse(&self, program: &mut backend::ast::Program, _: ()) -> Result<()> {
-        match *self {
-            webidl::ast::Interface::NonPartial(ref interface) => {
-                interface.webidl_parse(program, ())
+impl<'a, 'b> WebidlParse<&'a FirstPass<'b>> for webidl::ast::Includes {
+    fn webidl_parse(
+        &self,
+        program: &mut backend::ast::Program,
+        first_pass: &'a FirstPass<'b>,
+    ) -> Result<()> {
+        match first_pass.mixins.get(&self.includee) {
+            Some(mixin) => {
+                if let Some(non_partial) = mixin.non_partial {
+                    for member in &non_partial.members {
+                        member.webidl_parse(program, (&self.includer, first_pass))?;
+                    }
+                }
+                for partial in &mixin.partials {
+                    for member in &partial.members {
+                        member.webidl_parse(program, (&self.includer, first_pass))?;
+                    }
+                }
+            }
+            None => warn!("Tried to include missing mixin {}", self.includee),
+        }
+        Ok(())
+    }
+}
+
+impl<'a, 'b> WebidlParse<&'a FirstPass<'b>> for webidl::ast::Interface {
```

Comment:
I'd personally recommend avoiding the `WebidlParse` trait entirely as this looks like it may be getting a little too verbose, but otherwise perhaps the `FirstPass` parameter could be part of the trait signature to avoid the wordiness of lifetimes here?

Reply:
+1 for `FirstPass` as part of the trait signature, since pretty much every impl is using it.